### DISCUSSION TITLE
run a script in Discovery using run_discovery_script_file and get design back

### DISCUSSION THREAD

**Discussion by kmahajan-cadfem** [ansys/pyansys-geometry#1011] [General]
as far as I understand the `run_discovery_script_file` does not change design directly on the Python side because the default for `import_design` is set to `False`.

when a user is running a discovery script knowingly then that means the design should be changed on the python side as well, and hence I do not understand the default value of the `import_design` keyword.

can any one shed some light on it?
 
 
 as an example, following is an excerpt from one test

```python
from ansys.geometry.core import Modeler
from ansys.geometry.core.connection.backend import BackendType
from ansys.geometry.core.errors import GeometryRuntimeError
from ansys.geometry.core.math.point import Point2D
from ansys.geometry.core.sketch import Sketch
from ansys.geometry.core import launch_modeler_with_discovery
modeler_discovery = launch_modeler_with_discovery()

modeler = modeler_discovery

design = modeler.create_design("Integrated_Example")
design.extrude_sketch("Box", Sketch().box(Point2D([0, 0]), 1, 1), 1)
values = modeler.run_discovery_script_file(
    "integrated_script.py", {"radius": "1"}
)
# Script creates a 2nd body
assert int(values["numBodies"]) == 2
design = modeler.get_active_design()
assert len(design.bodies) == 2
```

content of the `integrated_script.py`

```python
doc = GetActiveWindow().Document
radius = MM(int(argsDict["radius"]))

sphere = Sphere.Create(Frame.World, radius)
u_range = Interval.Create(0, 2 * math.pi)
v_range = Interval.Create(-math.pi / 2, math.pi / 2)
box = BoxUV.Create(u_range, v_range)
b = Body.CreateSurfaceBody(sphere, box)
db = DesignBody.Create(doc.MainPart, "sphere", b)

numBodies = doc.MainPart.GetAllBodies().Count

result = {"numBodies": numBodies}
```

**Comment by RobPasMue** [comment]
Hi @kmahajan-cadfem! Thanks for opening this discussion. Couple of things from your script:

1. It seems like you have several unused import statements. I tried to simplify your script to the minimal reproducible example:

```py
from ansys.geometry.core.math.point import Point2D
from ansys.geometry.core.sketch import Sketch
from ansys.geometry.core import launch_modeler_with_discovery

# Launch discovery
modeler = launch_modeler_with_discovery()

# Create a new design
design = modeler.create_design("Integrated_Example")

# Sketch a box and extrude it
design.extrude_sketch("Box", Sketch().box(Point2D([0, 0]), 1, 1), 1)

# Run a scripting file that creates a second body
values = modeler.run_discovery_script_file("integrated_script.py", {"radius": "1"})

# Script creates a 2nd body
assert int(values["numBodies"]) == 2

# Get the active design
design = modeler.get_active_design()

# Verify number of bodies
assert len(design.bodies) == 2
```

2. You are right - ``import_design=True`` is the way you need to go in case you want to sync the remote design to your Python object. Otherwise it won't be synced properly

3. The misleading point or API is ``modeler.get_active_design()`` - this method does not sync the design on the backend. It just checks which is the design that is active in the ``Modeler`` object and returns it. This method is intended for products such as the Ansys Geometry Service and SpaceClaim that can handle multiple open documents (as of now). However, we are not syncing back the information from the backend. We are relying on the Python client's information for this so, if the ``design`` object has not been updated, it will not reflect the changes.

Conclusion: you are right, there is some unexpected behavior in your script - but it is due to the way ``design.get_active_design()`` works.

**Reply by RobPasMue** [reply]
Maybe the best solution is to implement an optional argument to that method - something like ``sync_with_remote : bool = False/True``.

Would you expect the sync method to be called automatically?

**Reply by kmahajan-cadfem** [reply]
@RobPasMue,
that would be great, to actually get the active and synced design back from the method, otherwise there would be many inconsistency problems that could arise in scripts.
as always thanks for prompt reply 🤘

**Reply by RobPasMue** [reply]
Okay great! Let me create an issue from this discussion to keep track of the implementation. It's my pleasure to support users needs! =)

By default, we will sync the design.

**Reply by RobPasMue** [reply]
Okay - feature merged in #1013 - it will be available in the next release we do =)

**Comment by RobPasMue** [comment]
Hi all!

From version ``0.4.11`` for ``ansys-geometry-core`` onwards, users can now request the active design of a ``Modeler`` instance and it will also sync with the backend (by default) to update any changes on the local ``Design`` object. This is done through the ``modeler.get_active_design()`` call.

For example, with the following script:

```py
from ansys.geometry.core.math.point import Point2D
from ansys.geometry.core.sketch import Sketch
from ansys.geometry.core import launch_modeler_with_discovery

# Launch discovery
modeler = launch_modeler_with_discovery()

# Create a new design
design = modeler.create_design("Integrated_Example")

# Sketch a box and extrude it
design.extrude_sketch("Box", Sketch().box(Point2D([0, 0]), 1, 1), 1)

# Run a scripting file that creates a second body
values = modeler.run_discovery_script_file("integrated_script.py", {"radius": "1"})

# Script creates a 2nd body
assert int(values["numBodies"]) == 2

# Verify number of bodies BEFORE syncing
assert len(design.bodies) == 1

# Get the active design
design = modeler.get_active_design()

# Verify number of bodies AFTER syncing
assert len(design.bodies) == 2
```

Another option is to call ``modeler.run_discovery_script(..., import_design=True)``--> This will sync the backend design automatically for you when executing a script.

Closing this discussion as it is now addressed.
